Makefile

```diff
@@ -108,42 +108,22 @@ uninstall:
 	pip freeze | grep -v "@" | xargs pip uninstall -y
 	rm -f $(env_tag) $(env_dev_tag) $(pre_deps_tag) $(install_tag)
 
-install: setup $(install_tag)
-
-install_dev: setup_dev $(install_tag)
-
-reinstall: install
-	@echo "==Rebuilding package distribution=="
-	${PYTHON} setup.py --quiet sdist
-	ls -rt  dist/* | tail -1 > dist/.build-tag
-	@echo "==Reinstalling package=="
-	${PYTHON} -m pip install --quiet $(shell ls -rt  dist/*.tar.gz | tail -1)
-	touch $(install_tag)
-
-reinstall_dev: install_dev
-	@echo "==Rebuilding package distribution=="
-	${PYTHON} setup.py --quiet sdist
-	ls -rt  dist/* | tail -1 > dist/.build-tag
-	@echo "==Reinstalling package=="
-	${PYTHON} -m pip install --quiet $(shell ls -rt  dist/*.tar.gz | tail -1)
-	touch $(install_tag)
-
-sort_imports: setup_dev
-	${PYTHON} -m isort $(folders)
+install: $(install_tag)
 
 format: setup_dev
 	${PYTHON} -m black $(folders)
+	${PYTHON} -m isort $(folders)
 
 lint: setup_dev
 	${PYTHON} -m flake8 $(folders)
 
-mypy: install_dev
+mypy: setup_dev $(install_tag)
 	${PYTHON} -m mypy --install-types --non-interactive --follow-imports silent $(folders)
 
-tests: install_dev
+tests: setup_dev $(install_tag)
 	${PYTHON} -m pytest tests
 
-checks: sort_imports format lint mypy tests
+checks: format lint mypy tests
```

Comment:
I'd remove format from this command since it does not actually run any check (it directly reformat code, without returning any failure).

Probably it would be better to place our re-formatting pipeline in a pre-commit hook so that it runs automatically before making any commit.

Reply:
Having said I would be fine with placing this in a pre-commit hook too, I don't think leaving format here too would do any harm. Before doing linting, static typing checks, and tests, I believe it makes sense to have the code in its best shape, with a standard, nice format.

Reply:
I agree it wouldn't do any harm but it wouldn't do anything useful either...
Now that I am thinking a bit about it I see quite some advantages in moving `make format` to a pre-commit hook:

- you make sure that code is automatically formatted correctly at every commit
- you make the `checks` command cleaner and a little quicker (relevant especially when used in workflows)

Reply:
As we have already discussed, I believe pre-hooks as well as make commands are ONLY tools that are provided to developers, to be used as they wish. One developer may prefer pre-hooks, while another may prefer NOT to have them on. We should keep an open mind on this. That's why I believe that leaving them in multiple places make really a lot of sense. We cannot enforce dev to neither use pre-commit nor use make commands. However, they should all comply with a standard CI pipeline

Reply:
Since code formatters do not return an error output when re-formatting some files, how do you plan to enforce file formatting without using pre-commit hooks?
CI workflows cannot do that and you cannot be sure that every possibile contributor will run `make format` before closing a PR...
Up to my knowledge, adding automatic formatting to a pre-commit hook is the only way to guarantee that it actually runs.

Reply:
we can always use black with the "--check" flag in the CI to make sure the code is nice and neat

When it comes to dev tools (such as make AND pre-commit), we can never be sure the user will use them. Also with pre-commit hooks one could either not set them or disable them (using the --no-verify flag). So both tools do not offer any guarantees strictly speaking. The only way to enforce them is via a CI pipeline, whereas the rests are tools to help developers to comply, but with no whatsoever strict enforcing.

Reply:
I suggests that in checks we make it dependent from format then. This should make everything flow like a charm. 

By the way: as said, I'm very in favour for git pre-commits, only leaving it to the discretion of the user. Do we have some recommended git-pre commit file as well as outline in the README the process to set that up? Could you @NicolaDonelli take care of it since you are definitely a master on this very topic?